Jan/staging

jan-kiszka opened the pull request:
* rework of the switchserialmode tool
* removal of the unused setledcolor script
* removal of Java build-dependency (yeah!)
* updated Yocto kernel to 4.4.60, including
    * fix for rs485 state reading
    * workaround for broken GPIO expander interrupts

jan-kiszka committed:
example: switchmode: Make coding style consistent and kernel-like

No functional changes.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: switchmode: Add support for bus termination

The IOT2000 allows to control RS485 and RS422 bus termination via
software. Add support for this to the mode switching tool.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: switchmode: Add support for reading back the mode

Allow to read out the current settings. Trivial, now that we have a
printing service.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: switchmode: Format usage output more nicely

Avoid overlong line, append a line break at the end, etc.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: switchmode: Return positive value on errors, no negative codes

Processes typically return positive codes on exit. Use 1 for errors
during execution and 2 for invalid parameters.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: switchmode: Avoid setting SER_RS485_RX_DURING_TX for RS232

Harmless - the driver will ignore it - but a bit confusing.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: switchmode: Factor out set_mode function

Prepares for dispatching to set vs. get mode later on.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: switchmode: Factor out print_mode

Will be reused by get_mode later on.

Note that the current patch queue for 4.4-yocto will not report the
correct mode back. To be fixed, and already fixed in the upstream queue.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
bsp: Select linux-yocto as kernel source

Recent versions of meta-intel choose their own kernel. But this is only
available for 4.9 onwards, while we are moving to 4.4-cip eventually.
Stay with linux-yocto with its 4.4 option to remain close.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: Remove Java dependency

The mraa recipe from iot-middleware comes with a mraa-java and upm-java
packages, but we do not need them and we do not install them. However,
the packages create build-time dependency on Java, which is prolonging
the build at best. At worst, it is importing build breakages due to
meta-java remaining fragile.

Remove this dependency by removing the mraa-java package

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
bsp: Update yocto kernel and patches

meta-intel's morty branch lags behind its pyro variant. So update the
source references manually to get latest 4.4.60 linux-yocto kernel.

Moreover, this fixes an issue in the Exar UART patch /wrt reading back
the rs485 state and adds the mfd patches that are pending for upstream.

Last but not least, this adds another workaround patch which addresses
the broken interrupts for the PCAL9535 I/O expander. Makes Arduino
shield reset work, e.g. The proper solution for this remains a firmware
update, though.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: switchmode: Add support for --help

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
switchmode: Move to bsp layer

This is an essential tool to control the board's X30 and X31 interfaces.
Therefore it should come with the BSP already, not only the example
image.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

stormc commented on meta-iot2000-example/recipes-tools/switchmode/files/switchmode.c:
```diff
@@ -21,12 +21,12 @@
-	printf("example: %s /dev/ttyS2 rs232", name);
+	printf("Usage: %s DEVICE MODE\n"
+	       "\n"
+	       "DEVICE\t\tThe device you for which you want to switch the mode.\n"
```
Hm, while at it, you may also correct the grammar: "[...] you for which you [...]"

stormc commented on meta-iot2000-example/recipes-tools/switchmode/files/switchmode.c:
```diff
@@ -42,13 +41,13 @@ int main(int argc, char *argv[])
 	if (argc != 3) {
 		print_usage(argv[0]);
-		return -EINVAL;
+		return 1;
```
Just nitpicking but an argument error should have a different exit code than a real error.

jan-kiszka replied:
Is that a convention? Is there one for the value as well?

stormc replied:
I'm not aware of a convention here but I'd like to be able to differentiate the two failure cases based on the exit code, e.g., in a script. As said, nitpicking... It's up to your taste.

stormc commented on meta-iot2000-example/recipes-tools/switchmode/files/switchmode.c:
```diff
@@ -59,6 +71,22 @@ static int set_mode(int file, char *device, char *mode)
+				fprintf(stderr,
+					"Termination not supported in RS232 "
+					"mode\n");
+				return 1;
```
This can be considered an argument error as well and consequently may have a different exit code.

stormc commented on meta-iot2000-example/recipes-tools/switchmode/files/switchmode.c:
```diff
@@ -59,6 +71,22 @@ static int set_mode(int file, char *device, char *mode)
+			rs485conf.flags |= SER_RS485_TERMINATE_BUS;
+		} else {
+			fprintf(stderr, "Invalid option \"%s\"\n", option);
+			return 1;
```
Same same.

stormc commented on meta-iot2000-example/recipes-tools/switchmode/files/switchmode.c:
```diff
@@ -122,6 +122,10 @@ int main(int argc, char *argv[])
 		print_usage(argv[0]);
 		return 1;
 	}
+	if (!strcmp(argv[1], "--help")) {
```
I think now it's time to use, e.g., `getopt`? :)

jan-kiszka replied:
I think it's on the edge. ;)

stormc reviewed (commented):
Minor changes.

jan-kiszka committed:
setledcolor: Move to example image

This does not belong to the bsp layer.

Reported-by: Claudius Heine <ch@denx.de>
Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: setledcolor: Fix dependency

This requires python-mraa to work.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: setledcolor: Fix setting of red LED

A variable name typo prevented that this path worked.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: setledcolor: Fix coding style

Make the script pep8-compliant.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka committed:
example: Install setledcolor

Otherwise, only few people will find and use it.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

jan-kiszka commented:
Updated switchmode according to feedback. I kept setledcolor on request by @SaschaWeisenberger, but I fixed it and install it now to the example image.

stormc commented on meta-iot2000-example/recipes-tools/setledcolor/files/setledcolor.py:
```diff
@@ -1,50 +1,48 @@
+        file.write("1")
+    else:
+        file.write("0")
+        file.close()
```
Well, this closes the file just when writing `0` to it. `file.close()` should be "de-indented".

jan-kiszka replied:
Good catch, thanks.

stormc commented on meta-iot2000-example/recipes-tools/setledcolor/files/setledcolor.py:
```diff
@@ -1,50 +1,48 @@
+    if not os.path.isfile(redLedFile):
+        print("Red LED not available")
+        return
+    file = open(redLedFile, 'w')
```
This may fail with an exception and the failure isn't handled here. Probably not worth the effort for this example script?

jan-kiszka replied:
Yes, but the most likely error - non-existing sysfs file - is caught above. OTH, one could avoid that check and handle the exception instead...
